### PR TITLE
Translation of BIR to LLVM IR 

### PR DESCRIPTION
**Pull request by shubhamnarlawar**
Purpose

Implementation of nballerina compiler.

Goals

Translation of Ballerina IR to LLVM IR.

Approach

Created a BIRReader class which reads BIR from the binary dump and store it in the object of BIRPackage class defined in BIR.h

Purpose of BIRReader class -

a. Reads Constant Pool entries.
b. Reads Module consisting of functions, basic blocks and instructions.

BIRPackage class consist of translate() function which translates functions, basic blocks and instructions to the LLVM IR.

### CODE REVIEW

**Comment by shubhamnarlawar** [comment]
@Kishanthan - Please review and suggest changes

File: BIR2llvmIR/BIR.h [hunk 1]
```diff
@@ -0,0 +1,707 @@
...
+    int     columnNum;
+
+  public:
+    Location();
+    Location(string name, int line, int col);
+    ~Location();
+
+    string getFileName();
+    int    getLineNum();
+    int    getColumnNum();
```
Comment (Kishanthan) [inline BIR2llvmIR/BIR.h:157]: As discussed in the slack channel, we need four values (lineStart, lineEnd, columnStart, columnEnd) and the source file name to define position information.

File: BIR2llvmIR/BIR.h [hunk 2]
```diff
@@ -0,0 +1,707 @@
...
+    Location * getLocation()                 { return loc; }
+    void       setLocation(Location *newLoc) { loc = newLoc; }
+
+    virtual void translate(LLVMModuleRef &modRef);
+};
+
+class PackageID {
+  private:
+    string        orgName;
+    string        sourceFileName;   // name is actually the source file name
```
Comment (Kishanthan) [inline BIR2llvmIR/BIR.h:180]: Shouldn't this be actually name? A packageId should have all the components (orgName + name + version)

File: BIR2llvmIR/BIR.h [hunk 3]
```diff
@@ -0,0 +1,707 @@
...
+    BIRNode (Location  *pos);
+    ~BIRNode ();
+
+    Location * getLocation()                 { return loc; }
+    void       setLocation(Location *newLoc) { loc = newLoc; }
+
+    virtual void translate(LLVMModuleRef &modRef);
+};
+
+class PackageID {
```
Comment (Kishanthan) [inline BIR2llvmIR/BIR.h:177]: Shall we not use the nameComp here? We can model the package with the following information 

- orgName
- name
- version
- sourceFileName

File: BIR2llvmIR/Makefile [hunk 4]
```diff
@@ -0,0 +1,95 @@
+# Define these parameters to reflect the proper directories.
```
Comment (manuranga) [inline BIR2llvmIR/Makefile:1]: Is this file committed accidentally ? shouldn't it be in the `build` dir?

File: BIR2llvmIR/BIR.h [hunk 5]
```diff
@@ -0,0 +1,707 @@
...
+    string getFirstNameComp()           { return nameComps.front(); }
+    string getLastNameComp()            { return nameComps.back(); }
+    bool   nameCompExists(string name);
+
+    typename list<string>::iterator getCompName(string name);  // Return the iterator to the entry matching name
+
+    virtual void translate(LLVMModuleRef &modRef);
+};
+
+class Symbol {
```
Comment (Kishanthan) [inline BIR2llvmIR/BIR.h:206]: We need to relook at whether we need Symbol, Scope, ScopeEntry and TypeSymbol classes here? The information they have is actually not needed for code-generation/BIR parsing.

File: BIR2llvmIR/driver.cpp [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+#include<iostream>
+#include<string>
+#include<fstream>
+#include<vector>
+using namespace std;
+
+int main()
+{
+  class BIRReader *Reader = new BIRReader("main-bir-dump");
+  class BIRPackage *BIRpackage = new BIRPackage ();
```
Comment (Kishanthan) [inline BIR2llvmIR/driver.cpp:12]: Do we have to new the BIRPackage here?

Comment (shubhamnarlawar) [reply BIR2llvmIR/driver.cpp:12]: Yes. BIRPackage object is passed to BIRReader which populates the data and returns it.

File: BIR2llvmIR/BIR.h (hunk 1 above)
Comment (shubhamnarlawar) [reply BIR2llvmIR/BIR.h:157]: I have added four values and source file name in Milestone 2 code. This PR is of Milestone 1. Would it be fine for now?

File: BIR2llvmIR/Makefile (hunk 4 above)
Comment (shubhamnarlawar) [reply BIR2llvmIR/Makefile:1]: The Makefile is present in the source code directory and hence generated files will present in the same directory and not the build directory. We will look into it and get back to you with appropriate way.

File: BIR2llvmIR/BIR.h (hunk 5 above)
Comment (shubhamnarlawar) [reply BIR2llvmIR/BIR.h:206]: Sure. I think `Symbol, Scope, ScopeEntry and TypeSymbol classes` are not required. Should I remove them?

File: BIR2llvmIR/BIR.h [hunk 7]
```diff
@@ -0,0 +1,707 @@
...
+
+  public:
+    AbstractInsn();
+    AbstractInsn(Location *pos, InstructionKind kind, Operand *lOp);
+    ~AbstractInsn();
+
+    InstructionKind getInstKind()   { return kind; }
+    Operand *       getLhsOperand() { return lhsOp; }
+
+    void          setFunction(BIRFunction *func) { BFunc = func; }
```
Comment (Kishanthan) [inline BIR2llvmIR/BIR.h:346]: Do we have to set the function ref here?

File: BIR2llvmIR/BIR.h [hunk 8]
```diff
@@ -0,0 +1,707 @@
...
+    Param *              getParam(int i)        { return requiredParams[i]; }
+    VarDecl *            getReceiver()          { return receiver; }
+    Param *              getRestParam()         { return restParam; }
+    vector<VarDecl *>    getLocalVars()         { return localVars; }
+    VarDecl *            getLocalVar(int i)     { return localVars[i]; }
+    VarDecl *            getReturnVar()         { return returnVar; }
+    vector<BasicBlockT *> getBasicBlocks()       { return basicBlocks; }
+    BasicBlockT *         getBasicBlock(int i)   { return basicBlocks[i]; }
+    string               getWorkerName()        { return workerName; }
+    LLVMBuilderRef       getLLVMBuilder()       { return builder; }
```
Comment (Kishanthan) [inline BIR2llvmIR/BIR.h:640]: why do we have to have the reference to the LLVM builder here?

File: BIR2llvmIR/BIR.h (hunk 2 above)
Comment (shubhamnarlawar) [reply BIR2llvmIR/BIR.h:180]: Actually, we are not storing any data to PackageID for now as we have - 

    a. orgName
    b. name
    c. version
    d. sourceFileName

in the BIRPackage class. Should I remove PackageID class?

File: BIR2llvmIR/driver.cpp (hunk 6 above)
Comment (manuranga) [reply BIR2llvmIR/driver.cpp:12]: if `deserialize` modifies `BIRpackage`, why does it has to return in back?

does following work?
```cpp
BIRPackage *BIRpackage = new BIRPackage ();
Reader->deserialize(BIRpackage);
```

File: BIR2llvmIR/BIR.h (hunk 7 above)
Comment (ramakotareddy) [reply BIR2llvmIR/BIR.h:346]: Yes Kishanthan. We need BIRFunction class object to call BIRFunction class Api's from other classes.

File: BIR2llvmIR/BIR.h (hunk 8 above)
Comment (ramakotareddy) [reply BIR2llvmIR/BIR.h:640]: Yes. We need LLVMBuilderRef to call the LLVM Translation Api's from each class. I have created and stored BuilderRef in BIRFunction class and accessing through BIRFunction class Object.

e.g: LLVMBuildAdd(LLVMBuilderRef, LLVMValueRef LHS, LLVMValueRef RHS,
                          const char *Name);

File: BIR2llvmIR/driver.cpp (hunk 6 above)
Comment (shubhamnarlawar) [reply BIR2llvmIR/driver.cpp:12]: Yes. This works for me. I will update it and send it with other necessary changes.

File: BIR2llvmIR/BIR.h (hunk 8 above)
Comment (manuranga) [reply BIR2llvmIR/BIR.h:640]: Suggestion here is, to keep the domain model (such as `BIRFunction`) separate from implementation classes (such as `LLVMBuilderRef `)

File: BIR2llvmIR/BIR.h (hunk 7 above)
Comment (manuranga) [reply BIR2llvmIR/BIR.h:346]: It's better if we can have the references going from top to bottom, (eg Module->Function->BB->Instruction) but not the other way around.

File: BIR2llvmIR/BIR.h (hunk 5 above)
Comment (manuranga) [reply BIR2llvmIR/BIR.h:206]: Will be addressed in #6

File: BIR2llvmIR/Makefile [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+      -lLLVMAMDGPUUtils -lLLVMAMDGPUInfo -lLLVMAArch64Disassembler \
+      -lLLVMMCDisassembler -lLLVMAArch64CodeGen -lLLVMCFGuard -lLLVMGlobalISel \
+      -lLLVMSelectionDAG -lLLVMAsmPrinter -lLLVMDebugInfoDWARF -lLLVMCodeGen \
+      -lLLVMTarget -lLLVMScalarOpts -lLLVMInstCombine -lLLVMAggressiveInstCombine \
+      -lLLVMTransformUtils -lLLVMBitWriter -lLLVMAnalysis -lLLVMProfileData \
+      -lLLVMObject -lLLVMTextAPI -lLLVMBitReader -lLLVMCore -lLLVMRemarks \
+      -lLLVMBitstreamReader -lLLVMAArch64AsmParser -lLLVMMCParser -lLLVMAArch64Desc \
+      -lLLVMMC -lLLVMDebugInfoCodeView -lLLVMDebugInfoMSF -lLLVMBinaryFormat \
+      -lLLVMAArch64Utils -lLLVMAArch64Info -lLLVMSupport -lLLVMDemangle -lpthread \
+      -ltinfo
```
Comment (manuranga) [inline BIR2llvmIR/Makefile:51]: ` -ltinfo` give a error while linking

File: BIR2llvmIR/driver.cpp (hunk 6 above)
Comment (shubhamnarlawar) [reply BIR2llvmIR/driver.cpp:12]: Will be addressed in #7

File: BIR2llvmIR/Makefile (hunk 9 above)
Comment (shubhamnarlawar) [reply BIR2llvmIR/Makefile:51]: Will be addressed in #8

File: BIR2llvmIR/BIR.h (hunk 8 above)
Comment (shubhamnarlawar) [reply BIR2llvmIR/BIR.h:640]: Will be addressed in #9

File: BIR2llvmIR/BIR.h (hunk 3 above)
Comment (shubhamnarlawar) [reply BIR2llvmIR/BIR.h:177]: Will be addressed in #6

File: BIR2llvmIR/BIR.h (hunk 2 above)
Comment (shubhamnarlawar) [reply BIR2llvmIR/BIR.h:180]: Will be addressed in #6

File: BIR2llvmIR/BIR.h (hunk 1 above)
Comment (shubhamnarlawar) [reply BIR2llvmIR/BIR.h:157]: Will be addressed in #10

File: BIR2llvmIR/BIR.h (hunk 7 above)
Comment (shubhamnarlawar) [reply BIR2llvmIR/BIR.h:346]: Will be addressed in #11

**Comment by shubhamnarlawar** [comment]
I have created issues for all the suggested changes above. The PR is ready to merge.

Please review and merge the PR.

**Comment by shubhamnarlawar** [comment]
PING.

Can you merge this PR so that we can send Milestone 2 PR?

**Review by Kishanthan — APPROVED** [review]

**Pull request merged**